tc/core/libraries.h

```diff
@@ -43,6 +43,13 @@ constexpr auto defines = R"C(
 #define inf __longlong_as_double(0x7ff0000000000000LL)
 )C";
 
+constexpr auto timestampFunction = R"C(
+__device__ unsigned long long __timestamp() {
+  unsigned long long startns;
+  asm volatile("mov.u64 %0,%%globaltimer;" : "=l"(startns));
```

Comment:
since you mention NVIDIA explicitly states we should not use globaltime, why not use something based on `clock64`, see an example [here](https://github.com/facebook/fbcuda/blob/8ef2af9b579b8610c59ab0ccf6e9075350de4320/Sleep.cu). We can get the frequency from the device property and turn that into time. This should also be significantly lower overhead than fectching data from global memory.